### PR TITLE
fix(script): decode CREATE tx arguments

### LINKED ISSUE #3053: `arguments` field for CREATE in broadcast log is `null`
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (20e7386 2022-08-31T00:09:48.876171Z)

### What command(s) is the bug in?

_No response_

### Operating System

macOS (Apple Silicon)

### Describe the bug

The `arguments` field is populated for function calls, but constructor arguments should be there as well, especially since they're the only way to distinguish multiple deploys of the same base contract.

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes #3053

also decode `arguments` for CREATE transactions

@mds1 mind giving this a try? 

## Solution
* move `TransactionWithMetadata` to its own module
* decode constructor/CREATE2call arguments if it's a create tx

### CODE REVIEW

**Comment by mds1** [comment]
I ran `forge init`, pasted the below into `Counter.s.sol`, and ran `forge script script/Counter.s.sol --tc Script0`. It succeeded but some reason I did not get a broadcast folder. 

```solidity
pragma solidity ^0.8.15;

import "forge-std/Script.sol";

contract A {
    address a;
    uint b;
    int c;
    bytes32 d;
    bool e;

  constructor(address _a, uint _b, int _c, bytes32 _d, bool _e) {
    a = _a;
    b = _b;
    c = _c;
    d = _d;
    e = _e;
  }
}

contract Script0 is Script {
  function run() external {
    vm.broadcast();
    new A(msg.sender, 2 ** 32, -1 * (2 ** 32), keccak256(abi.encode(1)), true);
  }
}
```

So then I started anvil and ran `forge script script/Counter.s.sol --tc Script0 --rpc-url http://127.0.0.1:8545` (and tried with `--broadcast` too), and got the below error (which says there should be a trace, but there isn't since I didn't pass a verbosity flag)

```
[⠰] Compiling...
No files changed, compilation skipped
Script ran successfully.
Error: 
Transaction failed when running the on-chain simulation. Check the trace above for more information.
```

So, not sure why these aren't working. Probably won't be able to look more until tomorrow, but basically I wanted to make sure all the different types were logged correctly (I probably should add `bytes` and `string` there too)

**Comment by mds1** [comment]
Ok, it does seem the arguments that are written are not encoded correctly. I tried running a script that had two address arguments of `0x794a61358D6845594F94dc1DB02A252b5b4814aD` and `0x4200000000000000000000000000000000000006`, but the output was

```json
"arguments": [
  "57600080fd5b506040516132ae3803806132ae83",
  "61016f565b6001600160a01b0382811660805281"
],
```

**Comment by mattsse** [comment]
I see, just need to encode them properly then, will write a test with your example, thanks!

File: cli/tests/it/script.rs [hunk 1]
```diff
@@ -613,3 +613,90 @@ forgetest_async!(test_custom_sender_balance, |prj: TestProject, cmd: TestCommand
...
+import "forge-std/Script.sol";
+
+contract A {
+    address a;
+    uint b;
+    int c;
+    bytes32 d;
+    bool e;
+
+  constructor(address _a, uint _b, int _c, bytes32 _d, bool _e) {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply cli/tests/it/script.rs:642]: extending the constructor arguments

**Comment by mattsse** [comment]
extended constructor args

**Comment by mds1** [comment]
Hmm, using `forge 0.2.0 (d7af7e3 2022-09-12T15:13:36.17375Z)`, just ran `forge script script/MyScript.s.sol --private-key $PRIVATE_KEY --rpc-url $OPTIMISM_RPC_URL --broadcast` and still got `"arguments": null,` even though the type is properly labeled CREATE

**Comment by mattsse** [comment]
hmm perhaps missing something else.

is this the example script?

**Comment by mds1** [comment]
That was a different one. I just setup a fresh forge project and ran this example script, and it does seem all arguments were logged properly:

Deploy script:
```solidity
function run() external {
  vm.broadcast();
  new A(msg.sender, 2 ** 32, -1 * (2 ** 32), keccak256(abi.encode(1)), true, hex"010203", "hello world");
}
```

Outputs:
```json
"arguments": [
  "0x00a329c0648769A73afAc7F9381E08FB43dBEA72",
  "4294967296",
  "-4294967296",
  "0xb10e2d527612073b26eecdfd717e6a320cf44b4afac2b0732d9fcbe2b7fa0cf6",
  "true",
  "0x010203",
  "hello world"
],
```

Not sure what I messed up yesterday (maybe looked at the wrong output file) but it does seem to be working!

However, the below gives me an error. Here is `src/A.sol`

```solidity
pragma solidity ^0.8.13;

contract A {
  address a;
  uint b;
  int c;
  bytes32 d;
  bool e;
  bytes f;
  string g;

  constructor(address _a, uint _b, int _c, bytes32 _d, bool _e, bytes memory _f, string memory _g) {
    a = _a;
    b = _b;
    c = _c;
    d = _d;
    e = _e;
    f = _f;
    g = _g;
  }
}

contract B {
  constructor(address _a, uint _b, int _c, bytes32 _d, bool _e, bytes memory _f, string memory _g) {
    new A(_a, _b, _c, _d, _e, _f, _g);
  }
}
```

Then in the script, I deploy `B`, which deploys `A` to test if it still shows the args for both A and B correctly. However I get the following output:

```
$ forge script script/Deploy.s.sol --rpc-url $OPTIMISM_RPC_URL -vvv
[⠆] Compiling...
No files changed, compilation skipped
Script ran successfully.
Error: 
Transaction failed when running the on-chain simulation. Check the trace above for more information.

Context:
- Failed to decode constructor arguments
```

**Comment by mattsse** [comment]
adding another test for this and investigate what's wrong here

**Comment by mattsse** [comment]
@mds1 fixed by finding the _last_ constructor in the bytecode

**Comment by mds1** [comment]
Ok that worked! I also tested it by changing contract B so that it deploys A in a method instead of a constructor, and that worked as expected. So this LGTM, thanks a lot @mattsse 🙌

**Pull request merged**